### PR TITLE
Expand documentation for data parallelism and distributed optimizer

### PR DESCRIPTION
**Pull request by timmoon10**
# What does this PR do ?

This expands the documentation for data parallelism and the distributed optimizer

@erhoo82 This may also be a good place to add FSDP documentation.

**Collection**: NLP

# Changelog 
- Expand documentation for data parallelism and the distributed optimizer

# Usage

n/a

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Matches style of https://github.com/NVIDIA/NeMo/pull/9184

### CODE REVIEW

**Comment by jgerh** [comment]
Hello, I assigned myself as a reviewer. I am the tech writer on the NeMo team and provided a review of https://github.com/NVIDIA/NeMo/pull/9184. I will make similar copyedits to the content. Thank you.

File: docs/source/features/parallelisms.rst [hunk 1]
```diff
@@ -5,15 +5,80 @@ Parallelisms
 
 NeMo Megatron supports 5 types of parallelisms (which can be mixed together arbitrarily):
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:6]: change "5" to "five"

File: docs/source/features/parallelisms.rst [hunk 2]
```diff
@@ -5,15 +5,80 @@ Parallelisms
...
+the model copies consistent with each other.
+
 Distributed Data Parallelism
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Distributed Data Parallelism (DDP) creates idential copies of the model across multiple GPUs.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Distributed Data Parallelism (DDP) keeps model copies consistent by
+synchronizing parameter gradients before each optimization step. More
+specifically it sums gradients over all model copies using an
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:22]: Change "More specifically it . . ." to "More specifically, it . . ."

File: docs/source/features/parallelisms.rst [hunk 3]
```diff
@@ -5,15 +5,80 @@ Parallelisms
...
+footprint on individual GPUs. It also has, in theory, the same
+communication volume as DDP and its communication pattern has more
+opportunities for overlapping with compute.
+
+Enable Data Parallelism
+~~~~~~~~~~~~~~~~~~~~~~~
+
+DDP is the default parallelism scheme when NeMo is run on multiple
+GPUs. Enabling other parallelism schemes in the model configuration
+will decrease the size of the DP group, i.e. the number of identical
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:50]: Change "will decrease the size of the DP group, i.e. the number of identical . . ." to "will decrease the size of the DP group, for example, the number of identical . . .

File: docs/source/features/parallelisms.rst [hunk 4]
```diff
@@ -5,15 +5,80 @@ Parallelisms
...
+model copies.
+
+In order to enable the distributed optimizer, set
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:53]: Change "In order to enable the . . ." to "To enable the . . ."

File: docs/source/features/parallelisms.rst [hunk 5]
```diff
@@ -5,15 +5,80 @@ Parallelisms
...
+``contiguous_param_buffer``  False      Allocate parameters as views into a large buffer. Helps avoid some data copies.
+``contiguous_grad_buffer``   True       Allocate parameter gradients as views into a large buffer. Helps avoid some data copies.
+===========================  =========  ==================================================================================================================================
+
+See the keyword arguments in `Apex DistributedFusedAdam <https://github.com/NVIDIA/apex/blob/master/apex/contrib/optimizers/distributed_fused_adam.py>`_ and `NeMo MegatronDistributedFusedAdam <https://github.com/NVIDIA/NeMo/blob/main/nemo/core/optim/distributed_adam.py>`_ for a full list of distributed optimizer options.
+
+Implement Data Parallelism
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+NeMo's DDP either uses PyTorch
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:74]: Change "Nemo's DDP either uses" to "The NeMo DDP uses either . . ."

Note: We avoid forming a possessive from a feature name, product name, or trademark. For example, we use NeMo rather than NeMo's or NVIDIA and not NVIDIA's.

File: docs/source/features/parallelisms.rst [hunk 6]
```diff
@@ -5,15 +5,80 @@ Parallelisms
...
+`DistributedDataParallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html>`_
+(default) or a custom implementation (if custom multi-precision
+training is enabled with ``megatron_amp_O2``).
+
+NeMo's distributed optimizer is built on top of
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:79]: Change "Nemo's distributed . . ." to "The Nemo distributed . . ."

**Review by jgerh — COMMENTED** [review]
I reviewed the changes and provided a few edits.

File: docs/source/features/parallelisms.rst [hunk 7]
```diff
@@ -3,17 +3,82 @@
 Parallelisms
 ------------
 
-NeMo Megatron supports 5 types of parallelisms (which can be mixed together arbitrarily):
+NeMo Megatron supports five types of parallelisms (which can be mixed together arbitrarily):
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:6]: Change to: NeMo Megatron supports five types of parallelisms (which can be mixed together arbitrarily).

Please delete colon and replace with a period

File: docs/source/features/parallelisms.rst [hunk 8]
```diff
@@ -3,17 +3,82 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+DDP in NeMo either uses PyTorch
+`DistributedDataParallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html>`_
+(default) or a custom implementation (if custom multi-precision
+training is enabled with ``megatron_amp_O2``).
+
+The distributed optimizer in NeMo is built on top of
+`DistributedFusedAdam <https://github.com/NVIDIA/apex/blob/master/apex/contrib/optimizers/distributed_fused_adam.py>`_
+from Apex.
```
Comment (jgerh) [inline docs/source/features/parallelisms.rst:81]: Comment for #86. Change **Tensor Parallelism (TP)** to Tensor Parallelism

Remove bold style.

**Review by jgerh — COMMENTED** [review]
I reviewed the file and added copyedits and formatting changes. Please review and implement the changes or ask me questions about the suggestions.

**Review by jgerh — APPROVED** [review]
